File: Cpp/src/json_device.cpp
/*
JsonTalkie - Json Talkie is intended for direct IoT communication.
Original Copyright (c) 2025 Rui Seixas Monteiro. All right reserved.
This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.
This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
Lesser General Public License for more details.
https://github.com/ruiseixasm/JsonTalkie
*/
#include "json_device.h"
#include <iostream>

JsonDevice::JsonDevice(std::shared_ptr<BroadcastSocket> socket, const std::string& name)
    : socket_(socket), name_(name.empty() ? "Device-" + device_id_.substr(0, 8) : name) {}

JsonDevice::~JsonDevice() {
    stop();
}

bool JsonDevice::start() {
    if (!socket_->open()) return false;
    running_ = true;
    thread_ = std::thread(&JsonDevice::listenLoop, this);
    return true;
}

void JsonDevice::stop() {
    running_ = false;
    if (thread_.joinable()) thread_.join();
    socket_->close();
}

bool JsonDevice::sendJson(const std::string& json) {
    return socket_->send(json);
}

void JsonDevice::setMessageCallback(MessageCallback callback) {
    callback_ = callback;
}

void JsonDevice::listenLoop() {
    while (running_) {
        auto received = socket_->receive();
        if (received && callback_) {
            callback_(received->first);
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(10));
    }
}

File: Python/json_device.py
import json
import threading
import uuid
from typing import Dict, Any

from broadcast_socket import BroadcastSocket

class JsonDevice:
    """Device with managed socket lifecycle."""

    def __init__(self, socket: BroadcastSocket, device_name: str = None):
        self._socket = socket  # Composition over inheritance
        self._device_id = str(uuid.uuid4())
        self._name = device_name or f"Device-{self._device_id[:8]}"
        self._running = False
    
    def start(self) -> bool:
        """Start message processing (no network knowledge)."""
        if not self._socket.open():
            return False
        self._running = True
        self._thread = threading.Thread(target=self._listen_loop, daemon=True)
        self._thread.start()
        return True
    
    def stop(self):
        """Stop processing (delegates cleanup to socket)."""
        self._running = False
        if hasattr(self, '_thread'):
            self._thread.join()
        self._socket.close()
    
    def _listen_loop(self):
        """Processes raw bytes from socket."""
        while self._running:
            received = self._socket.receive()
            if received:
                data, _ = received  # Explicitly ignore (ip, port)
                self._handle_message(data)
    
    def _handle_message(self, data: bytes):
        """Handles message content only."""
        try:
            message = json.loads(data.decode('utf-8'))
            self.on_message(message)  # No IP/port exposed!
        except (UnicodeDecodeError, json.JSONDecodeError) as e:
            print(f"[{self._name}] Invalid message: {e}")
    
    def send_json(self, message: Dict[str, Any]) -> bool:
        """Sends messages without network awareness."""
        return self._socket.send(json.dumps(message).encode('utf-8'))
    
    def on_message(self, message: Dict[str, Any]):
        """Override this to handle business logic."""
        print(f"[{self._name}] Received: {message}")

